File: Cyaim.WebSocketServer/Clients/cyaim-websocket-client-js/src/WebSocketClient.ts
import WebSocket from "ws";
import { encode, decode } from "@msgpack/msgpack";
import {
        WebSocketClientOptions,
        SerializationProtocol,
} from "./WebSocketClientOptions";

export interface MvcRequestScheme {
        id: string;
        target: string;
        body?: any;
}

export interface MvcResponseScheme {
        id: string;
        target: string;
        status: number;
        msg?: string;
        body?: any;
}

/**
 * Normalize a server JSON response (PascalCase: Status/Id/Msg/Body) to the client's lowercase shape.
 * Tolerant of camelCase too. / 将服务端 PascalCase 响应归一化为客户端小写形状（同时容忍 camelCase）。
 */
function normalizeResponse(r: any): MvcResponseScheme {
        return {
                id: r.Id ?? r.id,
                target: r.Target ?? r.target,
                status: r.Status ?? r.status ?? 0,
                msg: r.Msg ?? r.msg,
                body: r.Body ?? r.body,
        };
}

/**
 * WebSocket client for connecting to Cyaim.WebSocketServer
 * 用于连接到 Cyaim.WebSocketServer 的 WebSocket 客户端
 */
export class WebSocketClient {
        private serverUri: string;
        private channel: string;
        private options: WebSocketClientOptions;
        private webSocket: WebSocket | null = null;
        private pendingResponses: Map<
                string,
                {
                        resolve: (value: any) => void;
                        reject: (error: Error) => void;
                }
        > = new Map();

        /**
         * Constructor / 构造函数
         */
        constructor(
                serverUri: string,
                channel: string = "/ws",
                options?: WebSocketClientOptions
        ) {
                this.serverUri = serverUri;
                this.channel = channel;
                this.options =
                        options ||
                        ({
                                protocol: SerializationProtocol.Json,
                        } as WebSocketClientOptions);
        }

        /**
         * Connect to server / 连接到服务器
         */
        async connect(): Promise<void> {
                if (
                        this.webSocket &&
                        this.webSocket.readyState === WebSocket.OPEN
                ) {
                        return;
                }

                this.disconnect();

                const uri = `${this.serverUri.replace(/\/$/, "")}${
                        this.channel
                }`;
                this.webSocket = new WebSocket(uri);

                return new Promise((resolve, reject) => {
                        this.webSocket!.on("open", () => {
                                this.webSocket!.on(
                                        "message",
                                        (data: WebSocket.Data) => {
                                                this.handleMessage(data);
                                        }
                                );
                                resolve();
                        });

                        this.webSocket!.on("error", (error) => {
                                reject(error);
                        });
                });
        }

        /**
         * Send request and wait for response / 发送请求并等待响应
         */
        async sendRequest<TRequest, TResponse>(
                target: string,
                requestBody?: TRequest
        ): Promise<TResponse> {
                if (
                        !this.webSocket ||
                        this.webSocket.readyState !== WebSocket.OPEN
                ) {
                        throw new Error(
                                "WebSocket is not connected. Call connect() first."
                        );
                }

                const requestId = this.generateId();
                const request: MvcRequestScheme = {
                        id: requestId,
                        target,
                        body: requestBody,
                };

                return new Promise<TResponse>((resolve, reject) => {
                        this.pendingResponses.set(requestId, {
                                resolve,
                                reject,
                        });

                        if (
                                this.options.protocol ===
                                SerializationProtocol.MessagePack
                        ) {
                                // 服务端 MessagePackRequestScheme 使用整数 [Key(0)]Id/[Key(1)]Target/[Key(2)]Body，
                                // 即 MessagePack 数组格式，因此按 [id, target, body] 数组编码（而非命名字段的 map）。
                                // The server scheme uses integer [Key] (array format), so encode as [id, target, body].
                                const requestData = encode([
                                        request.id,
                                        request.target,
                                        request.body ?? null,
                                ]);
                                this.webSocket!.send(requestData);
                        } else {
                                // 使用 JSON 序列化
                                const requestJson = JSON.stringify(request);
                                this.webSocket!.send(requestJson);
                        }

                        // Timeout after 30 seconds
                        setTimeout(() => {
                                if (this.pendingResponses.has(requestId)) {
                                        this.pendingResponses.delete(requestId);
                                        reject(new Error("Request timeout"));
                                }
                        }, 30000);
                });
        }

        /**
         * Stream a large payload (e.g. a file) to a streaming endpoint ([WebSocket(Stream = true)]) without
         * buffering it all in memory. The payload is framed as a binary message:
         *   [4-byte magic "\0WSU"][4-byte big-endian header length N][UTF8 JSON header {id,target,meta}][raw bytes...]
         * The response is correlated by id, exactly like sendRequest.
         * 向流式端点上传大负载（如文件），负载不整体缓冲；响应按 id 关联，用法同 sendRequest。
         *
         * @param source A Uint8Array/Buffer, a Node Readable stream, or any (async) iterable of byte chunks.
         */
        async uploadStream<TResponse = any>(
                target: string,
                source:
                        | Uint8Array
                        | NodeJS.ReadableStream
                        | AsyncIterable<Uint8Array>
                        | Iterable<Uint8Array>,
                meta?: any,
                options?: { chunkSize?: number; timeoutMs?: number }
        ): Promise<TResponse> {
                if (
                        !this.webSocket ||
                        this.webSocket.readyState !== WebSocket.OPEN
                ) {
                        throw new Error(
                                "WebSocket is not connected. Call connect() first."
                        );
                }

                const requestId = this.generateId();
                const headerBytes = Buffer.from(
                        JSON.stringify({ id: requestId, target, meta: meta ?? null }),
                        "utf8"
                );
                const prefix = Buffer.alloc(8);
                // magic "\0WSU"
                prefix[0] = 0x00;
                prefix[1] = 0x57;
                prefix[2] = 0x53;
                prefix[3] = 0x55;
                prefix.writeUInt32BE(headerBytes.length, 4);
                const chunkSize = options?.chunkSize ?? 256 * 1024;
                const timeoutMs = options?.timeoutMs ?? 300000;

                const promise = new Promise<TResponse>((resolve, reject) => {
                        this.pendingResponses.set(requestId, { resolve, reject });
                        setTimeout(() => {
                                if (this.pendingResponses.has(requestId)) {
                                        this.pendingResponses.delete(requestId);
                                        reject(new Error("Upload timeout"));
                                }
                        }, timeoutMs);
                });

                try {
                        // frame 1: magic + header-length + header — start of the fragmented binary message
                        await this.sendFrame(
                                Buffer.concat([prefix, headerBytes]),
                                false
                        );
                        // payload frames (never buffers the whole payload)
                        for await (const chunk of WebSocketClient.chunkSource(
                                source,
                                chunkSize
                        )) {
                                if (chunk.length > 0) {
                                        await this.sendFrame(chunk, false);
                                }
                        }
                        // terminator: an empty final frame closes the message
                        await this.sendFrame(Buffer.alloc(0), true);
                } catch (err) {
                        this.pendingResponses.delete(requestId);
                        throw err;
                }
                return promise;
        }

        /** Send one binary frame; resolves once the frame is written (natural backpressure). */
        private sendFrame(data: Uint8Array, fin: boolean): Promise<void> {
                return new Promise((resolve, reject) => {
                        this.webSocket!.send(
                                data,
                                { binary: true, fin },
                                (err?: Error) => (err ? reject(err) : resolve())
                        );
                });
        }

        /** Normalize any supported source into byte chunks no larger than chunkSize. */
        private static async *chunkSource(
                source: any,
                chunkSize: number
        ): AsyncGenerator<Uint8Array> {
                const emit = function* (buf: Uint8Array): Generator<Uint8Array> {
                        for (let off = 0; off < buf.length; off += chunkSize) {
                                yield buf.subarray(
                                        off,
                                        Math.min(off + chunkSize, buf.length)
                                );
                        }
                };
                if (source instanceof Uint8Array) {
                        yield* emit(source);
                        return;
                }
                if (
                        source &&
                        (typeof source[Symbol.asyncIterator] === "function" ||
                                typeof source[Symbol.iterator] === "function")
                ) {
                        for await (const part of source as AsyncIterable<any>) {
                                const buf =
                                        part instanceof Uint8Array
                                                ? part
                                                : Buffer.from(part);
                                yield* emit(buf);
                        }
                        return;
                }
                throw new Error(
                        "Unsupported upload source: expected Uint8Array, Readable stream, or (async) iterable of bytes."
                );
        }

        /**
         * Handle incoming message / 处理接收到的消息
         */
        private handleMessage(data: WebSocket.Data): void {
                try {
                        let response: MvcResponseScheme;

                        if (
                                this.options.protocol ===
                                SerializationProtocol.MessagePack
                        ) {
                                // 处理二进制消息（MessagePack）
                                let buffer: Uint8Array;
                                if (
                                        data instanceof Buffer ||
                                        data instanceof Uint8Array
                                ) {
                                        buffer = data as Uint8Array;
                                } else if (Array.isArray(data)) {
                                        buffer = Buffer.concat(data);
                                } else if (data instanceof ArrayBuffer) {
                                        buffer = new Uint8Array(data);
                                } else {
                                        console.error(
                                                "Unexpected MessagePack data type:",
                                                typeof data
                                        );
                                        return;
                                }
                                // 服务端 MessagePackResponseScheme 使用整数 [Key]：
                                // [0]Status [1]Msg [2]RequestTime [3]CompleteTime [4]Id [5]Target [6]Body（数组格式）。
                                // The server response scheme uses integer [Key], decoded as an array.
                                const arr = decode(buffer) as any;
                                response = Array.isArray(arr)
                                        ? {
                                                  status: arr[0],
                                                  msg: arr[1],
                                                  id: arr[4],
                                                  target: arr[5],
                                                  body: arr[6],
                                          }
                                        : normalizeResponse(arr);
                        } else {
                                // 处理文本消息（JSON）。服务端响应为 PascalCase（Status/Id/Msg/Body）。
                                // JSON response is PascalCase; normalize to the client's lowercase shape.
                                const message = data.toString();
                                response = normalizeResponse(JSON.parse(message));
                        }

                        const pending = this.pendingResponses.get(response.id);
                        if (pending) {
                                this.pendingResponses.delete(response.id);

                                if (response.status !== 0) {
                                        pending.reject(
                                                new Error(
                                                        response.msg ||
                                                                "Unknown error"
                                                )
                                        );
                                } else {
                                        pending.resolve(response.body);
                                }
                        }
                } catch (error) {
                        console.error("Failed to parse response:", error);
                }
        }

        /**
         * Disconnect from server / 断开服务器连接
         */
        async disconnect(): Promise<void> {
                if (this.webSocket) {
                        this.webSocket.close();
                        this.webSocket = null;
                }
                this.pendingResponses.clear();
        }

        /**
         * Generate unique ID / 生成唯一 ID
         */
        private generateId(): string {
                return `${Date.now()}-${Math.random()
                        .toString(36)
                        .substr(2, 9)}`;
        }
}

File: Cyaim.WebSocketServer/Clients/cyaim-websocket-client-python/cyaim_websocket_client/websocket_client.py
import asyncio
import json
import uuid
from typing import Dict, Any, Optional, TypeVar
import websockets
import msgpack
from .websocket_client_options import WebSocketClientOptions, SerializationProtocol

T = TypeVar('T')


def _normalize_response(resp: Any) -> Dict[str, Any]:
    """Normalize a server response into {id, target, status, msg, body}.

    - JSON responses use PascalCase (Status/Id/Msg/Body).
    - MessagePack responses use integer [Key] and decode to a list
      [Status, Msg, RequestTime, CompleteTime, Id, Target, Body].
    将服务端响应归一化：JSON 为 PascalCase；MessagePack 为整数 [Key]（解码为列表）。
    """
    if isinstance(resp, (list, tuple)):
        return {
            "status": resp[0] if len(resp) > 0 else 0,
            "msg": resp[1] if len(resp) > 1 else None,
            "id": resp[4] if len(resp) > 4 else None,
            "target": resp[5] if len(resp) > 5 else None,
            "body": resp[6] if len(resp) > 6 else None,
        }
    if isinstance(resp, dict):
        def pick(*keys):
            for k in keys:
                if k in resp:
                    return resp[k]
            return None
        return {
            "status": pick("Status", "status", 0) or 0,
            "msg": pick("Msg", "msg"),
            "id": pick("Id", "id", 4),
            "target": pick("Target", "target", 5),
            "body": pick("Body", "body", 6),
        }
    return {"status": 0, "msg": None, "id": None, "target": None, "body": None}


class WebSocketClient:
    """WebSocket client for connecting to Cyaim.WebSocketServer"""

    def __init__(self, server_uri: str, channel: str = "/ws", options: Optional[WebSocketClientOptions] = None):
        self.server_uri = server_uri.rstrip('/')
        self.channel = channel
        self.options = options or WebSocketClientOptions()
        self.websocket = None
        self.pending_responses: Dict[str, asyncio.Future] = {}
        self._listen_task: Optional[asyncio.Task] = None

    async def connect(self):
        """Connect to server"""
        uri = f"{self.server_uri}{self.channel}"
        self.websocket = await websockets.connect(uri)
        self._listen_task = asyncio.create_task(self._listen())

    async def _listen(self):
        """Listen for incoming messages"""
        if not self.websocket:
            return
        try:
            async for message in self.websocket:
                await self._handle_message(message)
        except websockets.exceptions.ConnectionClosed:
            pass

    async def _handle_message(self, message):
        """Handle incoming message"""
        try:
            if isinstance(message, str):
                if self.options.protocol != SerializationProtocol.JSON:
                    return
                raw = json.loads(message)
            elif isinstance(message, (bytes, bytearray)):
                if self.options.protocol != SerializationProtocol.MESSAGEPACK:
                    return
                raw = msgpack.unpackb(bytes(message), raw=False)
            else:
                return

            response = _normalize_response(raw)
            request_id = response.get('id')
            if request_id is not None and request_id in self.pending_responses:
                future = self.pending_responses.pop(request_id)
                if not future.done():
                    future.set_result(response)
        except Exception as e:
            print(f"Failed to parse response: {e}")

    async def send_request(self, target: str, request_body: Optional[Any] = None) -> Any:
        """Send request and wait for response"""
        if not self.websocket:
            raise RuntimeError("WebSocket is not connected. Call connect() first.")

        request_id = str(uuid.uuid4())

        future: asyncio.Future = asyncio.get_event_loop().create_future()
        self.pending_responses[request_id] = future

        if self.options.protocol == SerializationProtocol.MESSAGEPACK:
            # Server MessagePackRequestScheme uses integer [Key(0)]Id/[Key(1)]Target/[Key(2)]Body
            # (array format), so encode as [id, target, body].
            request_bytes = msgpack.packb([request_id, target, request_body])
            await self.websocket.send(request_bytes)
        else:
            request = {"id": request_id, "target": target}
            if request_body is not None:
                request["body"] = request_body
            await self.websocket.send(json.dumps(request))

        try:
            response = await asyncio.wait_for(future, timeout=30.0)
            if response.get('status', 0) != 0:
                raise RuntimeError(response.get('msg') or 'Unknown error')
            return response.get('body')
        except asyncio.TimeoutError:
            self.pending_responses.pop(request_id, None)
            raise RuntimeError("Request timeout")
        finally:
            self.pending_responses.pop(request_id, None)

    async def upload_stream(self, target: str, source: Any, meta: Optional[Any] = None,
                            chunk_size: int = 256 * 1024, timeout: float = 300.0) -> Any:
        """Stream a large payload (e.g. a file) to a streaming endpoint ([WebSocket(Stream = True)]) without
        buffering it all in memory. Framed as a binary message:
        [4-byte magic \\x00WSU][4-byte big-endian header length][UTF8 JSON header {id,target,meta}][raw bytes...].
        The response is correlated by id, like send_request.
        向流式端点上传大负载（如文件），负载不整体缓冲；响应按 id 关联。
        `source` may be bytes, a file-like object (with .read), or a (sync/async) iterable of byte chunks.
        """
        if not self.websocket:
            raise RuntimeError("WebSocket is not connected. Call connect() first.")

        request_id = str(uuid.uuid4())
        header = json.dumps({"id": request_id, "target": target, "meta": meta}).encode("utf-8")
        prefix = b"\x00WSU" + len(header).to_bytes(4, "big")

        future: asyncio.Future = asyncio.get_event_loop().create_future()
        self.pending_responses[request_id] = future

        async def _frames():
            # frame 1: magic + header-length + header; then payload frames (never buffers the whole payload)
            yield prefix + header
            async for chunk in self._chunk_source(source, chunk_size):
                if chunk:
                    yield bytes(chunk)

        try:
            # websockets sends an (async) iterable as a single fragmented binary message
            await self.websocket.send(_frames())
            response = await asyncio.wait_for(future, timeout=timeout)
            if response.get('status', 0) != 0:
                raise RuntimeError(response.get('msg') or 'Unknown error')
            return response.get('body')
        except asyncio.TimeoutError:
            self.pending_responses.pop(request_id, None)
            raise RuntimeError("Upload timeout")
        finally:
            self.pending_responses.pop(request_id, None)

    @staticmethod
    async def _chunk_source(source: Any, chunk_size: int):
        """Normalize any supported upload source into byte chunks no larger than chunk_size."""
        if isinstance(source, (bytes, bytearray, memoryview)):
            data = bytes(source)
            for off in range(0, len(data), chunk_size):
                yield data[off:off + chunk_size]
            return
        read = getattr(source, "read", None)
        if callable(read):
            while True:
                chunk = read(chunk_size)
                if asyncio.iscoroutine(chunk):
                    chunk = await chunk
                if not chunk:
                    break
                yield chunk
            return
        if hasattr(source, "__aiter__"):
            async for part in source:
                data = bytes(part)
                for off in range(0, len(data), chunk_size):
                    yield data[off:off + chunk_size]
            return
        if hasattr(source, "__iter__"):
            for part in source:
                data = bytes(part)
                for off in range(0, len(data), chunk_size):
                    yield data[off:off + chunk_size]
            return
        raise TypeError("Unsupported upload source: expected bytes, a file-like object, or an iterable of bytes.")

    async def disconnect(self):
        """Disconnect from server"""
        if self._listen_task:
            self._listen_task.cancel()
            self._listen_task = None
        if self.websocket:
            await self.websocket.close()
            self.websocket = None
        self.pending_responses.clear()
